src/api/nbt_template.py

```diff
@@ -0,0 +1,333 @@
+                    continue
+
+                other_tag[tag_name] = template_tag.apply_to(nbt_tag)
+            elif template_tag.default_value is not None:
+                other_tag[tag_name] = template_tag.create_entry_from_template()
+
+        return other_tag
+
+    def create_entry_from_template(self, initial_value=None) -> NBTCompoundEntry:
+        if initial_value is None:
+            return NBTCompoundEntry(self.default_value)
+        return NBTCompoundEntry(initial_value)
+
+
+class NBTEntry:
+    """
+    Generic container used to represent a NBT entry that can have a template applied to it
+    """
+
+    __slots__ = ("tag_type", "value")
+
+    def __init__(self, tag_type: str, value: Any):
+        self.tag_type = tag_type
+        self.value = value
+
+    def __eq__(self, other: NBTEntry) -> bool:
+        return self.tag_type == other.tag_type and self.value == other.value
+
+    def __ne__(self, other: NBTEntry) -> bool:
+        return not self == other
+
+
+class NBTListEntry(UserList):
+    """
+    Container to represent a NBT List entry
+    """
+
+    tag_type = "list"
+
+    @property
+    def value(self):
+        return self.data
+
+
+class NBTCompoundEntry(UserDict):
+    """
+    Container to represent a NBT Compound entry
+    """
+
+    tag_type = "compound"
+
+    @property
+    def value(self):
+        return self.data
+
+
+class TemplateLoader:
+    """
+    Handles the loading of NBT template json files
+    """
+
+    def __init__(self, template_dir: str):
+        self._template_dir = template_dir
+
+        iter_chain = chain(
+            iglob(join(self._template_dir, "entities", "*.json")),
+            iglob(join(self._template_dir, "blockentities", "*.json")),
+            iglob(join(self._template_dir, "misc", "*.json")),
+        )
+        self._templates = {basename(p)[:-5]: p for p in iter_chain}
+        self._entity_template = self.load_template("entity")
+        self._item_template = self.load_template("item")
+
+    @property
+    def template_dir(self) -> str:
+        return self._template_dir
+
+    @property
+    def templates(self) -> Dict[str, str]:
+        return self._templates
+
+    def _tag_hook(self, json_object: dict) -> Union[dict, NBTStruct]:
+        if "tag_type" not in json_object:
+            return json_object
+
+        tag_type: str = json_object["tag_type"]
+
+        if tag_type.startswith("<") and tag_type.endswith(">"):
+            return self.load_template(tag_type[1:-1])
+
+        if tag_type == "compound":
+            structure = json_object["structure"]
+            if isinstance(structure, dict):
+                return NBTCompoundStructure(
+                    structure, default_value=json_object.get("default")
+                )
+            return NBTCompoundStructure(NBTStructure("any"))
+
+        if tag_type == "list":
+            return NBTListStructure(
+                json_object["schema"], default_value=json_object.get("default")
```

Comment:
Seems like the schema is either a string or a dict but the `NBTListStructure` is expecting a `NBTStructure`.
Am I missing something?

Reply:
When the JSON is being parsed, the value in `json_object["schema"]` has already been parsed by `_tag_hook()`, so when accessing it again with `json_object["schema"]` the value will now be some form of `NBTStructure`/`NBTCompoundStructure`/etc.

Reply:
Ok, fair enough. I wish there was a way to make it more obvious but I guess that's good enough